xcm/xcm-builder/src/integration_tests.rs

```diff
@@ -0,0 +1,282 @@
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: Parachain(other_para_id).into(),
+					},
+					// used to get a notification in case of success
+					Order::QueryHolding {
+						query_id,
+						dest: Parachain(PARA_ID).into(),
+						assets: vec![All],
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		let other_para_acc: AccountId = ParaId::from(other_para_id).into_account();
+		assert_eq!(Balances::free_balance(other_para_acc), amount);
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - 2 * amount);
+		assert_eq!(
+			mock::sent_xcm(),
+			vec![(
+				Parachain(PARA_ID).into(),
+				Xcm::QueryResponse {
+					query_id,
+					response: Response::Assets(vec![])
+				}
+			)]
+		);
+	});
+}
+
+/// Scenario:
+/// A parachain wants to move KSM from Kusama to Statemine.
+/// It withdraws funds and then teleports them to the destination.
+///
+/// This way of moving funds from a relay to a parachain will only work for trusted chains.
+/// Reserve based transfer should be used to move KSM to a community parachain.
+///
+/// Asserts that the balances are updated accordingly and the correct XCM is sent.
+#[test]
+fn teleport_to_statemine_works() {
+	use xcm::opaque::v0::prelude::*;
+	let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+	let balances = vec![(ALICE, INITIAL_BALANCE), (para_acc.clone(), INITIAL_BALANCE)];
+	kusama_like_with_balances(balances).execute_with(|| {
+		let statemine_id = 1000;
+		let other_para_id = 3000;
+		let amount =  10 * ExistentialDeposit::get();
+		let teleport_effects = vec![
+			buy_execution(5), // unchecked mock value
+			Order::DepositAsset { assets: vec![All], dest: X2(Parent, Parachain(PARA_ID)) },
+		];
+		let weight = 3 * BaseXcmWeight::get();
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::InitiateTeleport {
+						assets: vec![All],
+						dest: Parachain(other_para_id).into(),
+						effects: teleport_effects.clone(),
+					},
+				],
+			},
+			weight,
+		);
+		// teleports not allowed to community chains
+		assert_eq!(r, Outcome::Incomplete(weight, XcmError::UntrustedTeleportLocation));
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::InitiateTeleport {
+						assets: vec![All],
+						dest: Parachain(statemine_id).into(),
+						effects: teleport_effects.clone(),
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		// 2 * amount because of the incomplete failed attempt above
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - 2 * amount);
+		assert_eq!(
+			mock::sent_xcm(),
+			vec![(
```

Comment:
can't we also execute this and see its effects?

Reply:
Yeah, right now this test only tests for the first half (sending and burning) of teleportation, would like to also see it test the receiving part as well.